### PR TITLE
Ability to export bundle for deployment

### PR DESCRIPTION
**Pull request by jlc467**
Running `/packages/react-cosmos-webpack/bin/cosmos-export.js` will export an index.html and bundle.js that can be deployed, e.g. https://react-cosmos.github.io/?component=FlatrisGame&fixture=new-game-running

See https://github.com/react-cosmos/react-cosmos/issues/283 for background

Todo:
- [x] Tests will need to be updated/added

### CODE REVIEW

File: packages/react-cosmos-webpack/src/export.js [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+  const compiler = webpack(cosmosWebpackConfig);
+
+  compiler.run((err, stats) => {
+    if (err) {
+      console.error('Export Failed! See error below:');
+      console.error(err);
+    } else {
+      console.log(stats);
+      try {
+        fs.copySync(path.join(__dirname, 'static/favicon.ico'), `${cosmosWebpackConfig.output.path}/favicon.ico`);
```
Comment (skidding) [inline packages/react-cosmos-webpack/src/export.js]: This is a bit more tricky. Cosmos uses two index.htmls:
- [Index of the playground](https://github.com/react-cosmos/react-cosmos/blob/ed8ef3e09645cd69d0086339ea6e169294bd4aef/packages/react-cosmos-webpack/src/static/index.html), used to render nav and controls (this one is missing from the export at the moment)
- Index of the user, used for the contents of the loader frame (this one is generated by the export, but should be placed in the `loader/` dir, next to the `bundle.js` file)

Comment (jlc467) [inline packages/react-cosmos-webpack/src/export.js]: > Index of the user, used for the contents of the loader frame (this one is generated by the export, but should be placed in the loader/ dir, next to the bundle.js file)

Seems to work fine without this one, but then I use a web server that is redirecting to the index.html of the playground. Can you elaborate on when the `index of the user` comes into play?

**Comment by jlc467** [comment]
> I managed to tweak the URLs and make it work locally, but I'm still thinking of an elegant way to fix this. How is this working on your end?

The only way I could get it working is by redirecting everything to index.html with a web server. I tried by copying the export to my root and, while the playground shell would load correctly, the iframe would just show the directory browser which does have the `index of the user` in it. Sorry I can't be much help here.

**Comment by skidding** [comment]
Hey @jlc467, happy to see you're still working on this. Looking very good, let me know if you want me to review.

Regarding the other changes required for this to "just work", I started debugging on your branch and found all the missing pieces! Here are the changes that fixed every previous issue on my machine (tests not included):

- Removed comments about saving to memory, because we now save to disk as well
- Make paths relative
- Add production flag to minify/optimize bundle when exporting

image

- More relative paths

image

- More relative paths
- Disabled dev tools hook in production (export) because browser security yells at you when you try to access `window.parent` on the `file:///` protocol

image

Now, you have two options:
- We leave these for a separate PR to keep diffs smaller 
- You integrate these changes into this PR and ship the complete feature (home run)

Up to you. Also let me know if you need any more help. 

PS. This is going to make a great addition to the [upcoming 2.0 release](https://github.com/react-cosmos/react-cosmos/projects/3)!

**Comment by jlc467** [comment]
@skidding Thanks! This has been educational and fun to work on. I'll implement these changes sometime within the next few days, after which i think it would be good for you to review so we can wrap this PR up! 😎

**Comment by flaviusone** [comment]
@jlc467 are you planning to add anything else or is this ready? We are really excited and waiting for this feature!

**Comment by jlc467** [comment]
@flaviusone @skidding  Ready for code review

File: packages/react-cosmos-webpack/src/__tests__/entry.js [hunk 2]
```diff
@@ -43,7 +43,7 @@ describe('loader path', () => {
   let options;
 
   beforeAll(() => {
-    init('/loader/');
+    init('./loader/index.html');
```
Comment (skidding) [inline packages/react-cosmos-webpack/src/__tests__/entry.js]: This value mocks `window.location.pathname`, so it should be `/loader/index.html`. Tests should continue to pass.

File: packages/react-cosmos-webpack/src/__tests__/webpack-config.js [hunk 3]
```diff
@@ -225,3 +225,81 @@ describe('loaders', () => {
...
+    mockCosmosConfig = {
+      componentPaths: ['src/components'],
+      fixturePaths: ['test/fixtures'],
+      ignore: [],
+      globalImports: ['./global.css'],
+      hot: true,
+      outputPath: '__mock__outputPath'
+    };
+  });
+  describe('with shouldExport false', () => {
```
Comment (skidding) [inline packages/react-cosmos-webpack/src/__tests__/webpack-config.js]: It's annoying that I'm enforcing this without having an ESLint rule, but could you please add a newline between each `describe` and `test` blocks? The extra air improves readability.

**Comment by jlc467** [comment]
@skidding Changes made 😎

**Comment by skidding** [comment]
Congrats @jlc467! 🎉  🎆

**Pull request merged**